images/kcp-registrar/register.sh

```diff
@@ -78,65 +82,59 @@ get_clusters() {
     done
 }
 
 switch_org() {
-    KUBECONFIG=${kcp_kcfg} kubectl kcp workspace use ${KCP_ORG}
-    if ! (KUBECONFIG=${kcp_kcfg} kubectl api-resources >> /dev/null 2>&1); then
-        printf "%s is not a valid organization, wrong kubectl context in use or connectivity issue\n" ${KCP_ORG}
-	usage
-	exit 1
+    kcp_kubectl kcp workspace use "${KCP_ORG}"
+    if ! (kcp_kubectl api-resources >>/dev/null 2>&1); then
+        kcp_kubectl api-resources
+        exit_error "${KCP_ORG} is not a valid organization, wrong kubectl context in use or connectivity issue"
     fi
 }
 
 switch_ws() {
-    if (KUBECONFIG=${kcp_kcfg} kubectl get workspaces -o name | grep "${KCP_WORKSPACE}"); then
-        printf "use existing workspace\n"
-	KUBECONFIG=${kcp_kcfg} kubectl kcp workspace use ${KCP_WORKSPACE}
+    if (kcp_kubectl get workspaces -o name | grep -q "${KCP_WORKSPACE}"); then
+        printf "  - Use existing workspace\n"
+        kcp_kubectl kcp workspace use ${KCP_WORKSPACE}
 
     else
-       printf "creating workspace %s\n" "${KCP_WORKSPACE}"
-       KUBECONFIG=${kcp_kcfg} kubectl kcp workspace create "${KCP_WORKSPACE}" --enter
+        printf "  - Create workspace %s\n" "${KCP_WORKSPACE}"
+        kcp_kubectl kcp workspace create "${KCP_WORKSPACE}" --enter
     fi
 }
 
 register() {
-    printf "Getting the list of registered clusters\n"
-    existing_clusters=$(KUBECONFIG=${kcp_kcfg} kubectl get workloadclusters -o name)
-
     for i in "${!clusters[@]}"; do
-        printf "Processing cluster %s\n" "${clusters[$i]}"
-        if echo "${existing_clusters}" | grep "${clusters[$i]}"; then
-            printf "Cluster already registered\n"
-        else
-            printf "Registering cluster\n"
-            KUBECONFIG=${kcp_kcfg} kubectl kcp workload sync "${clusters[$i]}" \
-                 --syncer-image ghcr.io/kcp-dev/kcp/syncer:release-0.4 \
-		 --resources deployments.apps,services,ingresses.networking.k8s.io,conditions.tekton.dev,pipelines.tekton.dev,pipelineruns.tekton.dev,pipelineresources.tekton.dev,runs.tekton.dev,tasks.tekton.dev,taskruns.tekton.dev > /tmp/syncer-${clusters[$i]}.yaml
-            KUBECONFIG=${DATA_DIR}/credentials/kubeconfig/compute/${kubeconfigs[$i]} kubectl apply --context ${contexts[$i]} -f /tmp/syncer-${clusters[$i]}.yaml 
-        fi
+        cluster_name="$(echo "${clusters[$i]}" | cut -d: -f1)"
+        printf "  - %s\n" "$cluster_name"
+        syncer_manifest="/tmp/syncer-$cluster_name.yaml"
+        kcp_kubectl kcp workload sync "$cluster_name" \
+            --syncer-image ghcr.io/kcp-dev/kcp/syncer:release-0.4 \
+            --resources deployments.apps,services,ingresses.networking.k8s.io,conditions.tekton.dev,pipelines.tekton.dev,pipelineruns.tekton.dev,pipelineresources.tekton.dev,runs.tekton.dev,tasks.tekton.dev,taskruns.tekton.dev \
+            >"$syncer_manifest"
+        KUBECONFIG="$credentials_dir/kubeconfig/compute/${kubeconfigs[$i]}" kubectl apply --context ${contexts[$i]} -f "$syncer_manifest"
```

Comment:
There are implications in re-registering a cluster that has already been registered. Frankly I am not even sure that the kcp workload sync command would work if a workload cluster with the same cluster name already exists.
We can surely validate and refine our operational model regarding upgrades and handling of out-of-band changes but I don't think that this PR is the place for it.

Reply:
I did validate that executing a `kcp workload sync` did not generate a failure.